### PR TITLE
boards/nucleo-f303re: add ADC feature

### PR DESCRIPTION
**Pull request by hugueslarrive**
### Contribution description
Add ADC feature.

### Testing procedure
Untested.

### Issues/PRs references
Depends on PR #15119.

### CODE REVIEW

File: boards/nucleo-f303re/include/periph_conf.h [hunk 1]
```diff
@@ -32,6 +32,28 @@
 extern "C" {
 #endif
 
+/**
+ * @name    ADC configuration
+ *
+ * Note that we do not configure all ADC channels,
+ * and not in the STM32F334 order.  Instead, we
```
Comment (aabadie) [inline boards/nucleo-f303re/include/periph_conf.h:39]:
```suggestion
 * and not in the STM32F303 order.  Instead, we
```

**Review by aabadie — COMMENTED** [review]
Tested and this doesn't work. I'll check the board and cpu datasheet to see if there's a mistake somewhere.

**Comment by aabadie** [comment]
In fact, the slow channels are not working. The others 3 are working well.

File: boards/nucleo-f303re/include/periph_conf.h [hunk 2]
```diff
@@ -32,6 +32,28 @@
...
+ * just define 6 ADC channels, for the Nucleo
+ * Arduino header pins A0-A5
+ *
+ * @{
+ */
+static const adc_conf_t adc_config[] = {
+    { .pin = GPIO_PIN(PORT_A, 0), .dev = 0, .chan =  1 }, /* ADC1_IN1,  fast */
+    { .pin = GPIO_PIN(PORT_A, 1), .dev = 0, .chan =  2 }, /* ADC1_IN2,  fast */
+    { .pin = GPIO_PIN(PORT_A, 4), .dev = 1, .chan =  1 }, /* ADC2_IN1,  fast */
```
Comment (aabadie) [inline boards/nucleo-f303re/include/periph_conf.h:48]:
```suggestion
    { .pin = GPIO_PIN(PORT_A, 4), .dev = 1, .chan =  1 }, /* ADC12_IN1,  fast */
```

**Comment by aabadie** [comment]
Digging a bit more in the implementation and the peripheral driver is not able to manage ADC3 and ADC4. I have something that works when applying the following patch (might need a bit of cleanup...):

```diff
diff --git a/cpu/stm32/periph/adc_f3.c b/cpu/stm32/periph/adc_f3.c
index 18e1b11fbe..a7c79fa02a 100644
--- a/cpu/stm32/periph/adc_f3.c
+++ b/cpu/stm32/periph/adc_f3.c
@@ -35,25 +35,52 @@ static mutex_t locks[ADC_DEVS];
 
 static inline ADC_TypeDef *dev(adc_t line)
 {
+#ifdef ADC34_COMMON
+    if (adc_config[line].dev == 2) {
+        return (ADC_TypeDef *)(ADC3_BASE);
+    }
+    if (adc_config[line].dev == 3) {
+        return (ADC_TypeDef *)(ADC4_BASE);
+    }
+#endif
+
     return (ADC_TypeDef *)(ADC1_BASE + (adc_config[line].dev << 8));
 }
 
 static inline void prep(adc_t line)
 {
     mutex_lock(&locks[adc_config[line].dev]);
-/* Enable the clock here only if it will be disabled by done, else just
- * enable it once in adc_init() */
-#if defined(RCC_AHBENR_ADC1EN)
-    periph_clk_en(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+    /* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+     * enable the clock only once here. */
+    if (adc_config[line].dev <= 1) {
+#if defined(RCC_AHBENR_ADC12EN)
+        periph_clk_en(AHB, RCC_AHBENR_ADC12EN);
+#elif defined(RCC_AHBENR_ADC1EN)
+        periph_clk_en(AHB, RCC_AHBENR_ADC1EN);
+#endif
+    }
+#if defined(RCC_AHBENR_ADC34EN)
+    if (adc_config[line].dev >= 2) {
+        periph_clk_en(AHB, RCC_AHBENR_ADC34EN);
+    }
 #endif
 }
 
 static inline void done(adc_t line)
 {
-/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
- * don't disable the clock as the other device may still use it. */
-#if defined(RCC_AHBENR_ADC1EN)
-    periph_clk_dis(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+    /* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+     * disable the clock only once here. */
+    if (adc_config[line].dev <= 1) {
+#if defined(RCC_AHBENR_ADC12EN)
+        periph_clk_dis(AHB, RCC_AHBENR_ADC12EN);
+#elif defined(RCC_AHBENR_ADC1EN)
+        periph_clk_dis(AHB, RCC_AHBENR_ADC1EN);
+#endif
+    }
+#if defined(RCC_AHBENR_ADC34EN)
+    if (adc_config[line].dev >= 2) {
+        periph_clk_dis(AHB, RCC_AHBENR_ADC34EN);
+    }
 #endif
     mutex_unlock(&locks[adc_config[line].dev]);
 }
@@ -67,21 +94,31 @@ int adc_init(adc_t line)
 
     /* Lock device and enable its peripheral clock */
     prep(line);
-/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
- * enable the clock only once here. */
-#if !defined(RCC_AHBENR_ADC1EN)
-    periph_clk_en(AHB, (RCC_AHBENR_ADC12EN << adc_config[line].dev));
-#endif
 
     /* Setting ADC clock to HCLK/1 is only allowed if AHB clock
      * prescaler is 1 */
     if (!(RCC->CFGR & RCC_CFGR_HPRE_3)) {
         /* set ADC clock to HCLK/1 */
-        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
+        if (adc_config[line].dev <= 1) {
+            ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
+        }
+#ifdef ADC34_COMMON
+        if (adc_config[line].dev >= 2) {
+            ADC34_COMMON->CCR |= ADC_CCR_CKMODE_0;
+        }
+#endif
     }
     else {
-        /* set ADC clock to HCLK/2 otherwise */
-        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
+        if (adc_config[line].dev <= 1) {
+            /* set ADC clock to HCLK/2 otherwise */
+            ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
+        }
+#ifdef ADC34_COMMON
+        if (adc_config[line].dev >= 2) {
+            /* set ADC clock to HCLK/2 otherwise */
+            ADC34_COMMON->CCR |= ADC_CCR_CKMODE_1;
+        }
+#endif
     }
 
     /* Configure the pin */
```

**Comment by hugueslarrive** [comment]
> Digging a bit more in the implementation and the peripheral driver is not able to manage ADC3 and ADC4. I have something that works when applying the following patch (might need a bit of cleanup...):
> 
> ```diff
> diff --git a/cpu/stm32/periph/adc_f3.c b/cpu/stm32/periph/adc_f3.c
> index 18e1b11fbe..a7c79fa02a 100644
> --- a/cpu/stm32/periph/adc_f3.c
> +++ b/cpu/stm32/periph/adc_f3.c
> @@ -35,25 +35,52 @@ static mutex_t locks[ADC_DEVS];
>  
>  static inline ADC_TypeDef *dev(adc_t line)
>  {
> +#ifdef ADC34_COMMON
> +    if (adc_config[line].dev == 2) {
> +        return (ADC_TypeDef *)(ADC3_BASE);
> +    }
> +    if (adc_config[line].dev == 3) {
> +        return (ADC_TypeDef *)(ADC4_BASE);
> +    }
> +#endif
> +
>      return (ADC_TypeDef *)(ADC1_BASE + (adc_config[line].dev << 8));
>  }
>  
>  static inline void prep(adc_t line)
>  {
>      mutex_lock(&locks[adc_config[line].dev]);
> -/* Enable the clock here only if it will be disabled by done, else just
> - * enable it once in adc_init() */
> -#if defined(RCC_AHBENR_ADC1EN)
> -    periph_clk_en(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
> +    /* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
> +     * enable the clock only once here. */
> +    if (adc_config[line].dev <= 1) {
> +#if defined(RCC_AHBENR_ADC12EN)
> +        periph_clk_en(AHB, RCC_AHBENR_ADC12EN);
> +#elif defined(RCC_AHBENR_ADC1EN)
> +        periph_clk_en(AHB, RCC_AHBENR_ADC1EN);
> +#endif
> +    }
> +#if defined(RCC_AHBENR_ADC34EN)
> +    if (adc_config[line].dev >= 2) {
> +        periph_clk_en(AHB, RCC_AHBENR_ADC34EN);
> +    }
>  #endif
>  }
>  
>  static inline void done(adc_t line)
>  {
> -/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
> - * don't disable the clock as the other device may still use it. */
> -#if defined(RCC_AHBENR_ADC1EN)
> -    periph_clk_dis(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
> +    /* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
> +     * disable the clock only once here. */
> +    if (adc_config[line].dev <= 1) {
> +#if defined(RCC_AHBENR_ADC12EN)
> +        periph_clk_dis(AHB, RCC_AHBENR_ADC12EN);
> +#elif defined(RCC_AHBENR_ADC1EN)
> +        periph_clk_dis(AHB, RCC_AHBENR_ADC1EN);
> +#endif
> +    }
> +#if defined(RCC_AHBENR_ADC34EN)
> +    if (adc_config[line].dev >= 2) {
> +        periph_clk_dis(AHB, RCC_AHBENR_ADC34EN);
> +    }
>  #endif
>      mutex_unlock(&locks[adc_config[line].dev]);
>  }
> @@ -67,21 +94,31 @@ int adc_init(adc_t line)
>  
>      /* Lock device and enable its peripheral clock */
>      prep(line);
> -/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
> - * enable the clock only once here. */
> -#if !defined(RCC_AHBENR_ADC1EN)
> -    periph_clk_en(AHB, (RCC_AHBENR_ADC12EN << adc_config[line].dev));
> -#endif
>  
>      /* Setting ADC clock to HCLK/1 is only allowed if AHB clock
>       * prescaler is 1 */
>      if (!(RCC->CFGR & RCC_CFGR_HPRE_3)) {
>          /* set ADC clock to HCLK/1 */
> -        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
> +        if (adc_config[line].dev <= 1) {
> +            ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
> +        }
> +#ifdef ADC34_COMMON
> +        if (adc_config[line].dev >= 2) {
> +            ADC34_COMMON->CCR |= ADC_CCR_CKMODE_0;
> +        }
> +#endif
>      }
>      else {
> -        /* set ADC clock to HCLK/2 otherwise */
> -        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
> +        if (adc_config[line].dev <= 1) {
> +            /* set ADC clock to HCLK/2 otherwise */
> +            ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
> +        }
> +#ifdef ADC34_COMMON
> +        if (adc_config[line].dev >= 2) {
> +            /* set ADC clock to HCLK/2 otherwise */
> +            ADC34_COMMON->CCR |= ADC_CCR_CKMODE_1;
> +        }
> +#endif
>      }
>  
>      /* Configure the pin */
> ```

Mutex do not prevent simultaneous usage of dev 0 and 1 or dev 2 and 3 so I still think we should not disable the clock in done()...

**Comment by aabadie** [comment]
> Mutex do not prevent simultaneous usage of dev 0 and 1 or dev 2 and 3 so I still think we should not disable the clock in done()...

You are right, my suggestion was just given as an example, and as I said it needs some cleanup :)

**Comment by hugueslarrive** [comment]
> > Mutex do not prevent simultaneous usage of dev 0 and 1 or dev 2 and 3 so I still think we should not disable the clock in done()...
> 
> You are right, my suggestion was just given as an example, and as I said it needs some cleanup :)

Ok I will do another PR to fix the adc_f3 driver.

**Comment by aabadie** [comment]
#15119 was merged, please squash and rebase!

**Review by aabadie — APPROVED** [review]
Already tested with #15119 

ACK

**Comment by aabadie** [comment]
Wait @hugueslarrive this PR is not merged yet !

**Comment by aabadie** [comment]
Although you restored the branch, I can't reopen the PR. Are you able to do it on your side @hugueslarrive ?

**Comment by aabadie** [comment]
Nevermind, I could reopen it. Let's wait for Murdock now :)

**Pull request merged**